### ISSUE TITLE
Azure Databricks: Input Source operations connected to node Adam_Allreduce/HorovodAllgather_gradient_tape_model_embedding

### ISSUE THREAD

**Issue by chenyang1759** [horovod/horovod#2404]
**Environment:**
1. Framework: Tensorflow Keras
2. Framework version: 2.2.0
3. Horovod version: 0.19.1
4. MPI version:
5. CUDA version: 10.1 Update 2
6. NCCL version: 2.7.3
7. Python version:  3.7.5
8. Spark / PySpark version: 3.0.0
9. OS and version: Ubuntu 18.04.4 LTS
10. GCC version:
11. CMake version:

**Checklist:**
1. Did you search issues to find if somebody asked this question before? Yes
2. If your question is about hang, did you read [this doc] (https://github.com/horovod/horovod/blob/master/docs/running.rst)? Yes
3. If your question is about docker, did you read [this doc] (https://github.com/horovod/horovod/blob/master/docs/docker.rst)? No
4. Did you check if you question is answered in the [troubleshooting guide](https://github.com/horovod/horovod/blob/master/docs/troubleshooting.rst)? Yes

**Your Question**

I am currently trying to run Horovod Runner on Azure Databricks, with these specs
 Databricks Runtime 7.0 ML
Driver: Standard_NC12
Worker: Standard_NC12

I ran the sample [notebook ](https://docs.databricks.com/_static/notebooks/deep-learning/mnist-tensorflow-keras.html) and it worked on my cluster.

However, when I run my model training code, it seems to be training for the first few epochs. However, after a few epochs, it fails from the following exception.

```
[1,0]<stdout>:1023-053219-sects931-10-139-64-6:3263:3273 [0] NCCL INFO Channel 01 : 0[100000] -> 1[200000] via direct shared memory
[1,1]<stdout>:1023-053219-sects931-10-139-64-6:3264:3274 [1] NCCL INFO Channel 01 : 1[200000] -> 0[100000] via direct shared memory
[1,0]<stdout>:1023-053219-sects931-10-139-64-6:3263:3273 [0] NCCL INFO 2 coll channels, 2 p2p channels, 2 p2p channels per peer
[1,1]<stdout>:1023-053219-sects931-10-139-64-6:3264:3274 [1] NCCL INFO 2 coll channels, 2 p2p channels, 2 p2p channels per peer
[1,0]<stdout>:1023-053219-sects931-10-139-64-6:3263:3273 [0] NCCL INFO comm 0x7f90edf81170 rank 0 nranks 2 cudaDev 0 busId 100000 - Init COMPLETE
[1,1]<stdout>:1023-053219-sects931-10-139-64-6:3264:3274 [1] NCCL INFO comm 0x7f9198c00f40 rank 1 nranks 2 cudaDev 1 busId 200000 - Init COMPLETE
[1,0]<stdout>:1023-053219-sects931-10-139-64-6:3263:3273 [0] NCCL INFO Launch mode Parallel
[1,1]<stdout>:EPOCH 0 LOSS 0.6931361556053162 part 5
[1,0]<stdout>:1/1 - 1s - loss: 0.6945
[1,0]<stdout>:EPOCH 0 LOSS 0.6944578886032104 part 0
[1,1]<stdout>:begin negative sampling
[1,0]<stdout>:begin negative sampling
[1,1]<stdout>:appending negative samples
[1,0]<stdout>:appending negative samples
[1,1]<stdout>:training partition loaded
[1,1]<stdout>:There are 10742560 rows of data
[1,0]<stdout>:training partition loaded
[1,0]<stdout>:There are 10706885 rows of data
[1,1]<stdout>:EPOCH 0 LOSS 0.6641014814376831 part 6
[1,0]<stdout>:1/1 - 1s - loss: 0.6658
[1,0]<stdout>:EPOCH 0 LOSS 0.665772020816803 part 1
[1,1]<stdout>:begin negative sampling
[1,0]<stdout>:begin negative sampling
[1,1]<stdout>:appending negative samples
[1,0]<stdout>:appending negative samples
[1,1]<stdout>:training partition loaded
[1,1]<stdout>:There are 10655940 rows of data
[1,0]<stdout>:training partition loaded
[1,0]<stdout>:There are 10820185 rows of data
[1,1]<stdout>:EPOCH 0 LOSS 0.6473209261894226 part 7
[1,0]<stdout>:1/1 - 1s - loss: 0.6353
[1,0]<stdout>:EPOCH 0 LOSS 0.635341227054596 part 2
[1,1]<stdout>:begin negative sampling
[1,0]<stdout>:begin negative sampling
[1,1]<stdout>:appending negative samples
[1,0]<stdout>:appending negative samples
[1,1]<stdout>:training partition loaded
[1,1]<stdout>:There are 10777575 rows of data
[1,0]<stdout>:training partition loaded
[1,0]<stdout>:There are 10690585 rows of data
[1,1]<stdout>:EPOCH 0 LOSS 0.610869824886322 part 8
[1,0]<stdout>:1/1 - 1s - loss: 0.6197
[1,0]<stdout>:EPOCH 0 LOSS 0.6196755170822144 part 3
[1,0]<stdout>:begin negative sampling
[1,0]<stdout>:appending negative samples
[1,0]<stdout>:training partition loaded
[1,0]<stdout>:There are 10708845 rows of data
[1,0]<stderr>:2020-10-28 21:12:58.917943: W tensorflow/core/framework/op_kernel.cc:1753] OP_REQUIRES failed at mpi_ops.cc:376 : Unknown: Horovod has been shut down. This was caused by an exception on one of the ranks or an attempt to allreduce, allgather or broadcast a tensor after one of the ranks finished execution. If the shutdown was caused by an exception, you should see the exception in the log before the first shutdown message.
[1,0]<stderr>:2020-10-28 21:12:58.918953: W tensorflow/core/framework/op_kernel.cc:1753] OP_REQUIRES failed at mpi_ops.cc:316 : Unknown: Horovod has been shut down. This was caused by an exception on one of the ranks or an attempt to allreduce, allgather or broadcast a tensor after one of the ranks finished execution. If the shutdown was caused by an exception, you should see the exception in the log before the first shutdown message.
[1,0]<stderr>:2020-10-28 21:12:58.919167: W tensorflow/core/framework/op_kernel.cc:1753] OP_REQUIRES failed at mpi_ops.cc:316 : Unknown: Horovod has been shut down. This was caused by an exception on one of the ranks or an attempt to allreduce, allgather or broadcast a tensor after one of the ranks finished execution. If the shutdown was caused by an exception, you should see the exception in the log before the first shutdown message.
[1,0]<stderr>:2020-10-28 21:12:58.919334: W tensorflow/core/framework/op_kernel.cc:1753] OP_REQUIRES failed at mpi_ops.cc:316 : Unknown: Horovod has been shut down. This was caused by an exception on one of the ranks or an attempt to allreduce, allgather or broadcast a tensor after one of the ranks finished execution. If the shutdown was caused by an exception, you should see the exception in the log before the first shutdown message.
[1,0]<stderr>:2020-10-28 21:12:58.917959: W tensorflow/core/framework/op_kernel.cc:1753] OP_REQUIRES failed at mpi_ops.cc:376 : Unknown: Horovod has been shut down. This was caused by an exception on one of the ranks or an attempt to allreduce, allgather or broadcast a tensor after one of the ranks finished execution. If the shutdown was caused by an exception, you should see the exception in the log before the first shutdown message.
[1,0]<stderr>:2020-10-28 21:12:58.919388: W tensorflow/core/framework/op_kernel.cc:1753] OP_REQUIRES failed at mpi_ops.cc:316 : Unknown: Horovod has been shut down. This was caused by an exception on one of the ranks or an attempt to allreduce, allgather or broadcast a tensor after one of the ranks finished execution. If the shutdown was caused by an exception, you should see the exception in the log before the first shutdown message.
[1,0]<stderr>:Traceback (most recent call last):
[1,0]<stderr>:  File "<string>", line 1, in <module>
[1,0]<stderr>:  File "/databricks/python/lib/python3.7/site-packages/sparkdl/horovod/runner.py", line 222, in wrapped_main
[1,0]<stderr>:    return_value = main(**kwargs)
[1,0]<stderr>:  File "<command-291512615524496>", line 47, in train_hvd
[1,0]<stderr>:  File "/databricks/python/lib/python3.7/site-packages/tensorflow/python/keras/engine/training.py", line 66, in _method_wrapper
[1,0]<stderr>:    return method(self, *args, **kwargs)
[1,0]<stderr>:  File "/databricks/python/lib/python3.7/site-packages/tensorflow/python/keras/engine/training.py", line 848, in fit
[1,0]<stderr>:    tmp_logs = train_function(iterator)
[1,0]<stderr>:  File "/databricks/python/lib/python3.7/site-packages/tensorflow/python/eager/def_function.py", line 580, in __call__
[1,0]<stderr>:    result = self._call(*args, **kwds)
[1,0]<stderr>:  File "/databricks/python/lib/python3.7/site-packages/tensorflow/python/eager/def_function.py", line 611, in _call
[1,0]<stderr>:    return self._stateless_fn(*args, **kwds)  # pylint: disable=not-callable
[1,0]<stderr>:  File "/databricks/python/lib/python3.7/site-packages/tensorflow/python/eager/function.py", line 2420, in __call__
[1,0]<stderr>:    return graph_function._filtered_call(args, kwargs)  # pylint: disable=protected-access
[1,0]<stderr>:  File "/databricks/python/lib/python3.7/site-packages/tensorflow/python/eager/function.py", line 1665, in _filtered_call
[1,0]<stderr>:    self.captured_inputs)
[1,0]<stderr>:  File "/databricks/python/lib/python3.7/site-packages/tensorflow/python/eager/function.py", line 1746, in _call_flat
[1,0]<stderr>:    ctx, args, cancellation_manager=cancellation_manager))
[1,0]<stderr>:  File "/databricks/python/lib/python3.7/site-packages/tensorflow/python/eager/function.py", line 598, in call
[1,0]<stderr>:    ctx=ctx)[1,0]<stderr>:
[1,0]<stderr>:  File "/databricks/python/lib/python3.7/site-packages/tensorflow/python/eager/execute.py", line 60, in quick_execute
[1,0]<stderr>:    inputs, attrs, num_outputs)
[1,0]<stderr>:tensorflow.python.framework.errors_impl.UnknownError[1,0]<stderr>:: 2 root error(s) found.
[1,0]<stderr>:  (0) Unknown:  Horovod has been shut down. This was caused by an exception on one of the ranks or an attempt to allreduce, allgather or broadcast a tensor after one of the ranks finished execution. If the shutdown was caused by an exception, you should see the exception in the log before the first shutdown message.
[1,0]<stderr>:	 [[node Adam_Allreduce/HorovodAllgather_gradient_tape_model_embedding_embedding_lookup_Reshape_1_0 (defined at <string>:152) ]]
[1,0]<stderr>:	 [[GroupCrossDeviceControlEdges_0/Adam/Adam/Const/_123]]
[1,0]<stderr>:  (1) Unknown:  Horovod has been shut down. This was caused by an exception on one of the ranks or an attempt to allreduce, allgather or broadcast a tensor after one of the ranks finished execution. If the shutdown was caused by an exception, you should see the exception in the log before the first shutdown message.
[1,0]<stderr>:	 [[node Adam_Allreduce/HorovodAllgather_gradient_tape_model_embedding_embedding_lookup_Reshape_1_0 (defined at <string>:152) ]]
[1,0]<stderr>:0 successful operations.
[1,0]<stderr>:0 derived errors ignored. [Op:__inference_train_function_1790]
[1,0]<stderr>:
[1,0]<stderr>:Errors may have originated from an input operation.
[1,0]<stderr>:Input Source operations connected to node Adam_Allreduce/HorovodAllgather_gradient_tape_model_embedding_embedding_lookup_Reshape_1_0:
[1,0]<stderr>: gradient_tape/model/embedding/embedding_lookup/Reshape_1 (defined at <command-291512615524496>:47)
[1,0]<stderr>:
[1,0]<stderr>:Input Source operations connected to node Adam_Allreduce/HorovodAllgather_gradient_tape_model_embedding_embedding_lookup_Reshape_1_0:
[1,0]<stderr>: gradient_tape/model/embedding/embedding_lookup/Reshape_1 (defined at <command-291512615524496>:47)
[1,0]<stderr>:
[1,0]<stderr>:Function call stack:
[1,0]<stderr>:train_function -> train_function
```

I've tried scaling down the training data to 1000 rows, but the error still persists. Has anyone encountered this issue? Is there any additional configuration that I need on Databricks? Any help would really be appreciated! Thanks!

**Comment by chenyang1759** [comment]
My model code:
```
def build_NeuMF_model_context(num_users, num_items, day_unique, hour_unique, storeid_unique, mf_latent=10, layers=[10], regs_layers=[0], reg_mf=0):
  
  import tensorflow as tf

  from tensorflow.keras.layers import Embedding, Input, Dense, Reshape, Flatten, Dropout

  from tensorflow.keras import Model
  from tensorflow.keras import optimizers
  from tensorflow.keras.regularizers import l2
  from tensorflow.keras import initializers
  user_input = Input(shape=(1,), name='user_input')
  item_input = Input(shape=(1,), name = 'item_input')
  day_input = Input(shape=(1,), name = 'day_of_week_input')
  hour_input = Input(shape=(1,), name='hour_of_day_input')
  store_input = Input(shape=(1,), name='storeid_input')
  
  
  MF_Embedding_User = Embedding(input_dim = num_users, output_dim = mf_latent, name = 'mf_user_embedding',
                              embeddings_initializer = 'normal', embeddings_regularizer = l2(reg_mf), input_length=1)
  MF_Embedding_Item = Embedding(input_dim = num_items, output_dim = mf_latent, name = 'mf_item_embedding',
                              embeddings_initializer = 'normal', embeddings_regularizer = l2(reg_mf), input_length=1)
  
  MLP_Embedding_User = Embedding(input_dim=num_users, output_dim=layers[0]//2, 
                                 name='mlp_embedding_user', embeddings_initializer = 'normal', 
                                 embeddings_regularizer = l2(regs_layers[0]), input_length=1)
  MLP_Embedding_Item = Embedding(input_dim=num_items, output_dim=layers[0]//2, 
                                 name='mlp_embedding_item', embeddings_initializer = 'normal', 
                                 embeddings_regularizer = l2(regs_layers[0]), input_length=1)
  
  day_embedding_size = int(min((day_unique/2), 50 ))
  hour_embedding_size = int(min((hour_unique/2), 50 ))
  store_embedding_size = int(min((storeid_unique/2), 50 ))
  
  day_embedding = Embedding(input_dim=day_unique, output_dim=day_embedding_size)(day_input)
  hour_embedding = Embedding(input_dim=hour_unique, output_dim=hour_embedding_size)(hour_input)  
  storeid_embedding = Embedding(input_dim=storeid_unique, output_dim=store_embedding_size)(store_input)  
  
  #Context portion
  day_flatten = Flatten()(day_embedding)
  hour_flatten = Flatten()(hour_embedding)
  storeid_flatten = Flatten()(storeid_embedding)
  merged = tf.keras.layers.concatenate([day_flatten, hour_flatten])
  meta_vector = tf.keras.layers.concatenate([merged, storeid_flatten])
  meta_vector = Dense(10, activation='relu', name='meta_layer')(meta_vector)
  
  #Matrix Factorization portion
  mf_user_latent = Flatten()(MF_Embedding_User(user_input))
  mf_item_latent = Flatten()(MF_Embedding_Item(item_input))
  
  mf_vector = tf.keras.layers.multiply([mf_user_latent, mf_item_latent])
  
  #MLP portion
  mlp_user_latent = Flatten()(MLP_Embedding_User(user_input))
  mlp_item_latent = Flatten()(MLP_Embedding_Item(item_input))
  
  mlp_vector = tf.keras.layers.concatenate([mlp_user_latent, mlp_item_latent])
  for i in range(1, len(layers)):
    layer = Dense(layers[i], kernel_regularizer=l2(regs_layers[i]), activation='relu', name='layer{}'.format(i))
    mlp_vector = layer(mlp_vector)
    
  # Concatenate layers together
  
  predict_vector_no_meta = tf.keras.layers.concatenate([mf_vector, mlp_vector])
  predict_vector = tf.keras.layers.concatenate([predict_vector_no_meta, meta_vector])
  
  # Final prediction layer
  prediction = Dense(1, activation='sigmoid', kernel_initializer='lecun_uniform', name = 'prediction')(predict_vector)
  
  
  model = Model(inputs=[user_input, item_input, day_input, hour_input, store_input], 
            outputs=prediction)
  return model
```

The Horovod Training Code:
```
def train_hvd(learning_rate=0.01):
  # Tensorflow has given up on pickling. We need to explicitly import its modules inside workers
  from tensorflow.keras import backend as K
  from tensorflow.keras.models import Sequential
  import tensorflow as tf
  from tensorflow import keras
  import horovod.tensorflow.keras as hvd
  # Horovod: initialize Horovod.
  hvd.init()

  # Horovod: pin GPU to be used to process local rank (one GPU per process)
  gpus = tf.config.experimental.list_physical_devices('GPU')
  for gpu in gpus:
    tf.config.experimental.set_memory_growth(gpu, True)

  if gpus:
    tf.config.experimental.set_visible_devices(gpus[hvd.local_rank()], 'GPU')

  
  print("model to be built")
  model = build_NeuMF_model_context(NUM_USERS, NUM_ITEMS, NUM_DAYS, NUM_HOURS, NUM_STORES)
  optimizer = optimizers.Adam(lr=learning_rate * hvd.size())
  print("model built")
  # Horovod: add Horovod Distributed Optimizer.
  optimizer = hvd.DistributedOptimizer(optimizer)

  model.compile(optimizer=optimizer,
                 loss='binary_crossentropy')

  print("model compiled")
  callbacks = [
      # Horovod: broadcast initial variable states from rank 0 to all other processes.
      # This is necessary to ensure consistent initialization of all workers when
      # training is started with random weights or restored from a checkpoint.
      hvd.callbacks.BroadcastGlobalVariablesCallback(0),
  ]
  
  if hvd.local_rank() == 0:
    callbacks.append(keras.callbacks.ModelCheckpoint(checkpoint_dir + '/checkpoint-{epoch}.ckpt', save_weights_only = True))

  epochs = 1
  for ep in range(epochs):
    print("Epoch {}".format(ep))
    if hvd.local_rank() == 0:
      for part in range(0, 5):
        (users, items, days, stores, hours, ratings) = get_dataset(part)
        hist = model.fit([users, items, days, hours, stores], ratings, callbacks=callbacks, batch_size=1000, epochs=1, shuffle=False, verbose=2, use_multiprocessing=False)
        print( "EPOCH {} LOSS {} part {}".format(ep, hist.history['loss'][0], part))
        gc.collect()
    else:
      for part in range(5, 9):
        (users, items, days, stores, hours, ratings) = get_dataset(part)
        hist = model.fit([users, items, days, hours, stores], ratings, callbacks=callbacks, batch_size=1000, epochs=1, shuffle=False, verbose=0, use_multiprocessing=False)
        print( "EPOCH {} LOSS {} part {}".format(ep, hist.history['loss'][0], part))
        gc.collect()
  
from sparkdl import HorovodRunner

hr = HorovodRunner(np=2)
hr.run(train_hvd, learning_rate=0.01)

```

**Comment by tgaddair** [comment]
Hey @chenyang1759, it looks like you're using Horovod v0.19.1 with TensorFlow 2 and running into issues with allgather gradient.  This looks consistent with #2110.  We addressed this in #2121, which was released with Horovod 0.20.  Is it possible to upgrade?

**Comment by chenyang1759** [comment]
Thanks for getting back to me so quickly! I'll give that a shot.
